### ISSUE TITLE
Internal error due to trailing whitespace

### ISSUE THREAD

**Issue by fprasx** [rust-lang/rustfmt#5462]
Hi, 

I recently received this error:
```
error[internal]: left behind trailing whitespace
  --> [Secret SuperClassified Directory]/rust/phenotype/phenotype-macro/src/lib.rs:90:90:94
   |
90 |                     let struct_field_placeholders = (0..unnamed.len()).map(syn::Index::from); 
   |                                                                                              ^
   |

warning: rustfmt has failed to format. See previous 1 errors.
```

The offending piece of code is:
```rust
    89                  syn::Fields::Unnamed( FieldsUnnamed { unnamed , ..}) => {
    90                      let struct_field_placeholders = (0..unnamed.len()).map(syn::Index::from); 
    91                      quote! {
    92                          #tag => {
    93                              // Safe because the tag guarantees that we are reading the correct field
    94                              let data = ::std::mem::ManuallyDrop::<#struct_name>::into_inner(unsafe { value.unwrap().#var_ident });
    95                              #ident::#var_ident ( #(data.#struct_field_placeholders),* )
    96                          }
    97                      }
    98                  }
```
Line 90 has one space of trailing whitespace.

I wasn't able to produce a minimum working example, but the file (and codebase) is public on my github.

File: [lib.rs](https://github.com/fprasx/peapod/blob/main/phenotype-macro/src/lib.rs)

Version info:
```
cargo
cargo 1.64.0-nightly (b1dd22e66 2022-07-09)

rustc
rustc 1.64.0-nightly (c396bb3b8 2022-07-10)

rustfmt
rustfmt 1.5.1-nightly (c396bb3b 2022-07-10)
```

**Comment by ytmimi** [comment]
Thanks for reaching out. Are you using any configuration options?

**Comment by fprasx** [comment]
Nope, just `cargo fmt`

**Comment by ytmimi** [comment]
Did some digging and found that there's a few things coming together here.

TLDR;
1) rustfmt doesn't know how to properly format the `quote!` macro #3406
2) rustfmt gives up on formatting a chain when any reformatting fails #3863
3) Since we gave up on formatting the chain we left behind whitespace that already existed in the source code.

And if you're curious here's a longer explanation:

As you can see by the screenshot, the source code contains a line with trailing whitespace.
<img width="952" alt="Screen Shot 2022-07-22 at 3 39 25 PM" src="https://user-images.githubusercontent.com/29028348/180515593-013dd869-2599-4ba4-a328-5b289ede9768.png">


Here's a modified, minimal code snippet that can be used to reproduce the issue:

```rust
fn foo() {
    {
        arms.push(
            match &var.fields {
                syn::Fields::Unnamed( FieldsUnnamed { unnamed , ..}) => {
                    let struct_field_placeholders = (0..unnamed.len()).map(syn::Index::from); 
                    quote! {
                        #tag => {
                            // Safe because the tag guarantees that we are reading the correct field
                            let data = ::std::mem::ManuallyDrop::<#struct_name>::into_inner(unsafe { value.unwrap().#var_ident });
                            #ident::#var_ident ( #(data.#struct_field_placeholders),* )
                        }
                    }
                }
            }
        )
    }
}
```

This is essentially a duplicate of #3863 that's leading to a `left behind trailing whitespace` error. If rustfmt is unable to format the code inside a chain'd method call `e.g arms.push`, then it refuses to rewrite any part of the chain. 

If you run rustfmt with the code snippet that I posted above you should get two errors: 1 for the trailing whitespace issue, and another for the `exceeded maximum width` error.

```
error[internal]: left behind trailing whitespace
 --> /rustfmt/issue_5462.rs:6:6:94
  |
6 |                     let struct_field_placeholders = (0..unnamed.len()).map(syn::Index::from); 
  |                                                                                              ^
  |

error[internal]: line formatted, but exceeded maximum width (maximum: 100 (see `max_width` option), found: 130)
  --> /rustfmt/issue_5462.rs:10:10:101
   |
10 |                             let data = ::std::mem::ManuallyDrop::<#struct_name>::into_inner(unsafe { value.unwrap().#var_ident });
   |                                                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |

warning: rustfmt has failed to format. See previous 2 errors.
```

In this case the `max_width` issue is preventing us from reformatting the chain, and therefore we leave the trailing whitespace as it was written in the source code.

If you're wondering why we can't format the code to meet the `max_width` constraint, it essentially comes down to the fact that rustfmt is unable to format the `quote!` macro call  (relevant issue #3406). Long story short, formatting macro calls is hard because they're allowed to contain invalid rust syntax `e.g. #struct_name`, and rustfmt is meant to format valid rust syntax.

<hr>

Possible solutions include (but are not limited to):

1) bumping up the max_width to a large enough value. Looking at the error message from before `(maximum: 100 (see `max_width` option), found: 130)`, we see that the offending line had a width of 130. bumping the `max_width` to 130+ will allow rust to format the code

2) manually formatting the quote lines so that they fit within the max_width constraint and then running rustfmt.

**Comment by ytmimi** [comment]
Going to close as I don't think there's any immediate action the team can take to provide a fix for the specific confluence of issues we're running into here
